feat/add get transaction endpoint

affan880 opened the pull request.

jackli921 committed:
Add POST /transactions endpoint

- added transaction related schema

- added related service and controller logic

- added related unit tests

- added Decimal.js to package.json to handle monetary data accurately

affan880 committed:
Add GET /transactions/:id endpoint

- Implemented getTransactions controller to fetch transactions by user ID
- Added getTransactionsSchema for input validation
- Updated transaction service to include getTransactions method
- Modified transaction routes to include the new GET endpoint

jackli921 commented on apps/server/src/services/transaction.service.ts:
```diff
@@ -0,0 +1,23 @@
+  async createTransaction(
+    transactionData: createTransactionInput,
+  ): Promise<Transaction> {
+    return await prisma.transactions.create({
```
looks like this is for the code i added too, but I think we can remove the `await` from the this context.

Apparently, `await` just unwraps the Promise result so you can use it as a value, but if you're just returning it, there's no need to unwrap it only to rewrap it in another Promise again

jackli921 commented on apps/server/src/services/transaction.service.ts:
```diff
@@ -0,0 +1,23 @@
+  },
+
+  async getTransactions(id: number): Promise<Transaction[]> {
+    return await prisma.transactions.findMany({
```
Same here as above.

jackli921 commented on apps/server/src/services/transaction.service.ts:
```diff
@@ -0,0 +1,23 @@
+  async getTransactions(id: number): Promise<Transaction[]> {
+    return await prisma.transactions.findMany({
+      where: {
+        user_id: id,
```
could make client's life easier but asking database to return the collection in a sorted order by `created_at` and break ties using `id`.

```typescript
orderBy: [
  { created_at: 'desc' },
  { id: 'desc' } // fallback if timestamps are equal
]
```

jackli921 replied:
any follow up on the suggestion?

affan880 replied:
I'm sorry I didn't get this.

jackli921 replied:
you can add the condition to ask supabse to return the ccollection sorted in descending order by created_at date and then by id to break the tie if two or more transactions are created at the same time

affan880 replied:
I was doing the same thing previously I think ?

jackli921 committed:
Add GET /categories Endpoint

- added categoryController, categoryService, categoryRouter

- added related unit tests

jackli921 committed:
Add requireAuth to protected API paths

- added requireAuth middleware to protected endpoints

- added mock requireAuth middleware to bypass it for controller unit tests

kayoMichael commented on apps/server/src/controllers/transaction.controller.ts:
```diff
@@ -0,0 +1,66 @@
+  const validationResult = createTransactionSchema.safeParse(req.body);
+  try {
+    if (!validationResult.success) {
+      if (env.NODE_ENV !== 'production') {
```
We should use the env schema for NODE_ENV

jackli921 replied:
This came from my PR, which is already merged in. Affan can rebase onto development and make the change as a part of this PR?

jackli921 replied:
could we rebase to reflect this change on this PR?

affan880 replied:
Hi, I rebased the PR

kayoMichael commented on apps/server/src/schemas/transactionSchema.ts:
```diff
@@ -0,0 +1,21 @@
+import Decimal from 'decimal.js';
+import { z } from 'zod';
```
One thing I noticed is that we are inconsistently using `zod` or `zod/v4` in our current code base. I think for consistency within the entire project we should stick to `zod`, so can you change all current occurences of `zod/v4` to `zod`

jackli921 replied:
could we resolve this?

kayoMichael reviewed (changes requested):
Good work. Are we doing PATCH and Delete API endpoints in another PR?

jackli921 commented:
Remember to rebase onto development as new PRs have been merged.

affan880 committed:
Remove unnecessary await and add sorting to getTransactions

affan880 committed:
Refactor zod import and remove redundant sorting in transaction service

kayoMichael commented on apps/server/src/controllers/transaction.controller.ts:
```diff
@@ -0,0 +1,66 @@
+};
+
+const getTransactions = async (
+  req: Request,
```
Aren't we not having a transaction table page. Then we will definitely need a GET ALL API and pagination on the API although it doesn't need to be implemented in this PR if you don't want to.

jackli921 replied:
What is Michael's question? We do have a transaction table/list page right?

jackli921 commented on apps/server/src/routes/transaction.routes.ts:
```diff
@@ -0,0 +1,13 @@
+const transactionRouter = Router();
+
+transactionRouter.post('/', createTransaction);
+transactionRouter.get('/:id', getTransactions);
```
The `/transaction/:id` endpoint should fetch one trasaction with a particular transactionId. But I think you're trying to get all the transaction of a particular user, then `/transactions/user/:id` is a more accurate in describing what the endpoint does.

affan880 replied:
I have updated the endpoint

affan880 committed:
- Changed GET endpoint from /transactions/:id to /transactions/user/:id for clarity and specificity in fetching transactions by user ID.
